### PR TITLE
OSNet Model Node

### PR DESCRIPTION
**Pull request by jovi-s**
New model node for person re-identification using OSNet.

New model node: JDE

- [x] PDR can be viewed [here](https://www.notion.so/e4d1884e5284462fb2db8c033d3369fe?v=1afeb0c8264645dbbd05ad108dc4d3e9)
- [x] CDR can be viewed [here](https://www.notion.so/31075a7ba3ad4e0c9418fd5e193438b4?v=4533ca95d43a4723afed6696d481a7b7)

### CODE REVIEW

File: peekingduck/pipeline/nodes/model/osnetv1/osnet_files/osnet.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+
+    def __init__(  # pylint: disable=too-many-arguments, unused-argument
+        self,
+        num_classes: int,
+        blocks: List[OSBlock],
+        layers: List[int],
+        channels: List[int],
+        feature_dim: int = 512,
+        loss: str = "softmax",
+        IN: bool = False,
```
Comment (liyier90) [inline peekingduck/pipeline/nodes/model/osnetv1/osnet_files/osnet.py:134]: Similar to above this shouldn't be all caps

File: peekingduck/pipeline/nodes/model/osnetv1/osnet_files/osnet_ain.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+
+class OSBlockINin(nn.Module):
+    """Omni-scale feature learning block with instance normalization."""
+
+    def __init__(  # pylint: disable=unused-argument
+        self,
+        in_channels: int,
+        out_channels: int,
+        reduction: int = 4,
+        T: int = 4,
```
Comment (liyier90) [inline peekingduck/pipeline/nodes/model/osnetv1/osnet_files/osnet_ain.py:124]: This shouldn't be all caps, also try to reference the paper if possible to use a self-descriptive name

File: peekingduck/pipeline/nodes/model/osnetv1/osnet_files/osnet_ain.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+    def __init__(  # pylint: disable=too-many-arguments, unused-argument
+        self,
+        num_classes: int,
+        blocks: List[OSBlock],
+        layers: List[int],
+        channels: List[int],
+        feature_dim: int = 512,
+        loss: str = "softmax",
+        conv1_IN: bool = False,
```
Comment (liyier90) [inline peekingduck/pipeline/nodes/model/osnetv1/osnet_files/osnet_ain.py:171]: Shouldn't contain all caps

File: peekingduck/pipeline/nodes/model/osnetv1/osnet_model.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+from peekingduck.pipeline.nodes.model.osnetv1.osnet_files.feature_extractor import (
+    FeatureExtractor,
+)
+
+
+class OSNetModel:  # pylint: disable=too-few-public-methods
+    """OSNet re-ID model for person identification."""
+
+    def __init__(self, config: Dict[str, Any]) -> None:
+        super().__init__()
```
Comment (liyier90) [inline peekingduck/pipeline/nodes/model/osnetv1/osnet_model.py:59]: I think this is not needed

File: peekingduck/pipeline/nodes/model/osnetv1/osnet_model.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        self.extractor = FeatureExtractor(
+            model_name=model_names[config["model_type"]],
+            model_path=str(model_path),
+            device=config["device"],
+        )
+        self.logger.info(
+            f"Model name: {model_names[config['model_type']]}\n"
+            " - params: 2,193,616\n"
+            " - flops: 978,878,352"
+        )
```
Comment (liyier90) [inline peekingduck/pipeline/nodes/model/osnetv1/osnet_model.py:102]: Are there any other model config related values to be printed here? I see there's `multi_threshold` in the config file, should that be printed here with a proper description of what it stands for?

**Review by liyier90 — CHANGES_REQUESTED** [review]
Just some elaboration on some of the arguments and variables and change the path to no be restricted to Windows only.

File: peekingduck/configs/model/osnet.yml [hunk 6]
```diff
@@ -0,0 +1,17 @@
+input: ["img", "bboxes"]
+output: ["bboxes", "obj_tags"]
+
+device: "cpu"
+model_type: "osnet"
+query_root_dir: "C:/Users/../../data/input/reid"
```
[the remark this answers is not in the document]
Comment (jovi-s) [reply peekingduck/configs/model/osnet.yml:6]: Resolved to `<root_dir>/data/input/reid`

File: peekingduck/pipeline/nodes/model/osnet.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+    def run(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
+        """Function that reads the image input and returns the matching
+        bbox of the specified query object to be detected.
+
+        Args:
+            inputs (Dict): Dict of inputs with keys "img", "bboxes".
+
+        Returns:
+            outputs (Dict): Dict output of matching bbox with label.
```
[the remark this answers is not in the document]
Comment (jovi-s) [reply peekingduck/pipeline/nodes/model/osnet.py:78]: Elaborated.

File: peekingduck/pipeline/nodes/model/osnetv1/osnet_files/build_model.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+    Args:
+        name (str): Model name.
+        num_classes (int): Number of training identities.
+        loss (str): Loss function to optimize the model. Currently
+            supports "softmax" and "triplet". Default is "softmax".
+        use_gpu (bool): Whether to use gpu. Default is True.
+
+    Returns:
+        nn.Module
```
[the remark this answers is not in the document]
Comment (jovi-s) [reply peekingduck/pipeline/nodes/model/osnetv1/osnet_files/build_model.py:69]: Elaborated.

File: peekingduck/pipeline/nodes/model/osnetv1/osnet_files/feature_extractor.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    """
+
+    # pylint: disable=dangerous-default-value, too-many-arguments
+    def __init__(
+        self,
+        model_name: str = "",
+        model_path: str = "",
+        image_size: Tuple[int, int] = (256, 128),
+        pixel_mean: List = [0.485, 0.456, 0.406],
+        pixel_std: List = [0.229, 0.224, 0.225],
```
[the remark this answers is not in the document]
Comment (jovi-s) [reply peekingduck/pipeline/nodes/model/osnetv1/osnet_files/feature_extractor.py:76]: Added

File: peekingduck/pipeline/nodes/model/osnetv1/osnet_files/network_blocks.py [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        in_channels: int,
+        out_channels: int,
+        kernel_size: int,
+        stride: int = 1,
+        padding: int = 0,
+        groups: int = 1,
+        IN: bool = False,
```
[the remark this answers is not in the document]
Comment (jovi-s) [reply peekingduck/pipeline/nodes/model/osnetv1/osnet_files/network_blocks.py:59]: Renamed

File: peekingduck/pipeline/nodes/model/osnetv1/osnet_files/osnet.py (hunk 1 above)
Comment (jovi-s) [reply peekingduck/pipeline/nodes/model/osnetv1/osnet_files/osnet.py:134]: Renamed

File: peekingduck/pipeline/nodes/model/osnetv1/osnet_files/osnet_ain.py (hunk 2 above)
Comment (jovi-s) [reply peekingduck/pipeline/nodes/model/osnetv1/osnet_files/osnet_ain.py:124]: Changed to `num_streams`

File: peekingduck/pipeline/nodes/model/osnetv1/osnet_files/osnet_ain.py (hunk 3 above)
Comment (jovi-s) [reply peekingduck/pipeline/nodes/model/osnetv1/osnet_files/osnet_ain.py:171]: Changed to `conv1_instance_norm`

File: peekingduck/pipeline/nodes/model/osnetv1/osnet_model.py (hunk 5 above)
Comment (jovi-s) [reply peekingduck/pipeline/nodes/model/osnetv1/osnet_model.py:102]: Updated

File: peekingduck/pipeline/nodes/model/osnetv1/osnet_files/build_model.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+"""
+Build OSNet model.
+"""
+
+from __future__ import absolute_import
```
[the remark this answers is not in the document]
Comment (jovi-s) [reply peekingduck/pipeline/nodes/model/osnetv1/osnet_files/build_model.py:39]: Removed

File: peekingduck/pipeline/nodes/model/osnetv1/osnet_files/distance.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+"""
+Compute distance between query and gallery features.
+"""
+
+from __future__ import division, print_function, absolute_import
```
[the remark this answers is not in the document]
Comment (jovi-s) [reply peekingduck/pipeline/nodes/model/osnetv1/osnet_files/distance.py:39]: Removed

File: peekingduck/pipeline/nodes/model/osnetv1/osnet_model.py (hunk 4 above)
Comment (jovi-s) [reply peekingduck/pipeline/nodes/model/osnetv1/osnet_model.py:59]: Removed

**Pull request merged**